### PR TITLE
Unit testing in .NET

### PR DESCRIPTION
**Pull request by mikhailshilkov**
This is .NET SDK work for #2287: an API to unit-test stack deployments.

Introduces `Deployment.TestAsync<TStack>(IMocks mocks)` API to run in-memory mock-backed unit tests for any `TStack` stack class.

### CODE REVIEW

File: sdk/dotnet/Pulumi/Testing/Mocks.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        public Task<global::Pulumirpc.InvokeResponse> InvokeAsync(global::Pulumirpc.InvokeRequest request)
+            => throw new NotImplementedException();
+
+        public Task<global::Pulumirpc.ReadResourceResponse> ReadResourceAsync(
+            global::Pulumirpc.ReadResourceRequest request)
+            => throw new NotImplementedException();
+
+        public Task<global::Pulumirpc.RegisterResourceResponse> RegisterResourceAsync(
+            global::Pulumirpc.RegisterResourceRequest request)
+            => throw new NotImplementedException();
```
Comment (mikhailshilkov) [inline sdk/dotnet/Pulumi/Testing/Mocks.cs]: @CyrusNajmabadi I would appreciate your help in implementing the above three methods to pass calls `IMocks`

Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/Mocks.cs]: I took a stab on that, resolving this question

**Comment by mikhailshilkov** [comment]
Usage example in C#: https://gist.github.com/mikhailshilkov/e2f341d7d15174cea4d1041f5422d2ad

Note that the `MustNameSecurityGroup` test doesn't work currently: the security group array is always empty. Might be something about the changes in PR.

**Comment by lonegunmanb** [comment]
Hello @mikhailshilkov ,
I really like the style you demonstrated in [testswithresources](https://gist.github.com/mikhailshilkov/e2f341d7d15174cea4d1041f5422d2ad#file-testswithresources-cs) !

File: sdk/dotnet/Pulumi/Testing/Mocks.cs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        /// </summary>
+        /// <param name="message">Error message.</param>
+        Task HandleErrorAsync(string message);
+
+        /// <summary>
+        /// Invoked when a resource registers its outputs.
+        /// </summary>
+        /// <param name="urn">Resource URN.</param>
+        /// <param name="outputs">Outputs property bag.</param>
+        Task ResourceOutputsAsync(string urn, ImmutableDictionary<string, object> outputs);
```
Comment (pgavlin) [inline sdk/dotnet/Pulumi/Testing/Mocks.cs]: I don't think that this is necessary--this is essentially an implementation detail of the engine.

File: sdk/dotnet/Pulumi/Testing/Mocks.cs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        /// </summary>
+        /// <param name="resource">Resource instance.</param>
+        /// <param name="type">Resource type name.</param>
+        /// <param name="name">Resource name.</param>
+        /// <param name="inputs">Dictionary of resource input properties.</param>
+        /// <param name="provider">Provider.</param>
+        /// <param name="id">Resource identifier.</param>
+        /// <returns>A tuple of a resource identifier and resource state. State can be either a POCO
+        /// or a dictionary bag.</returns>
+        Task<(string id, object state)> NewResourceAsync(Resource resource, string type, string name,
```
Comment (pgavlin) [inline sdk/dotnet/Pulumi/Testing/Mocks.cs]: I don't think that passing the resource in is particularly safe: the resource won't really be ready until after this call has returned.

File: sdk/dotnet/Pulumi/Testing/Mocks.cs (hunk 2 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/Mocks.cs]: The point is to get the actual outputs of a stack in https://github.com/pulumi/pulumi/pull/3696/files#diff-c87e9a22bf55d4ae8e2b14e128465573R39
I agree this may be a bit too low-level for the user code. We can hide it from the public API surface.

Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/Mocks.cs]: Removed

File: sdk/dotnet/Pulumi/Testing/Mocks.cs (hunk 3 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/Mocks.cs]: Indeed, removed.

File: sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs [hunk 4]
```diff
@@ -43,7 +43,10 @@ public void RegisterTask(string description, Task task)
             {
                 lock (_taskToDescription)
                 {
-                    _taskToDescription.Add(task, description);
+                    if (!_taskToDescription.ContainsKey(task))
+                    {
+                        _taskToDescription.Add(task, description);
+                    }
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs]: why was this changed?  this indicates that we're adding hte task multiple times.  that's likely not a good thing.

1. if this is correct, we need to doc when/why this happens.
2. if this is correct, it's simpler to just do `_taskToDescription[task] = description`.

Note: the `.Add` was very intentional on my part to catch bugs.  i.e. i wanted it to crash if we somehow double registered a task.

File: sdk/dotnet/Pulumi/Deployment/Deployment_RootResource.cs [hunk 5]
```diff
@@ -39,12 +39,12 @@ public partial class Deployment
         private async Task<string> SetRootResourceWorkerAsync(Stack stack)
         {
             var resUrn = await stack.Urn.GetValueAsync().ConfigureAwait(false);
-            await this.Engine.SetRootResourceAsync(new SetRootResourceRequest
+            await this.Monitor.SetRootResourceAsync(new SetRootResourceRequest
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Deployment/Deployment_RootResource.cs]: i'm not a huge fan of directly exposing the monitor.  can we not go through a fake engine which itself can go through its monitor?

File: sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
+using System.Threading.Tasks;
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs]: copyrights.

File: sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+using Grpc.Core;
+using Pulumirpc;
+
+namespace Pulumi
+{
+    class GrpcMonitor : IMonitor
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs]:
```suggestion
    internal class GrpcMonitor : IMonitor
```

File: sdk/dotnet/Pulumi/Deployment/Deployment_RootResource.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply sdk/dotnet/Pulumi/Deployment/Deployment_RootResource.cs]: ignore the above.

Comment (CyrusNajmabadi) [reply sdk/dotnet/Pulumi/Deployment/Deployment_RootResource.cs]: I am still trying to reconcile this.  i don't love the incongruity here.  Our system has an 'Engine' and a 'Monitor' for different purposes.  It reads very weird to me to see this as normally a ResourceMonitorClient would not be setting the root resource.

File: sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs]: i'd almost prefer this as just having a generic `IRpc` interface which encapsulates all the rpc operations we need the deployment to perform.  Then we have our `GrpcRpc` impl of that interface.  That would read a lot better to me than calling this a `Monitor` which then talks to an `Engine`.

File: sdk/dotnet/Pulumi/Deployment/Deployment_RootResource.cs (hunk 5 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/Deployment_RootResource.cs]: Alright... is it a matter of finding a better name? Or do we have to keep two separate entities and mock both?

File: sdk/dotnet/Pulumi/Deployment/IMonitor.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
+using System.Threading.Tasks;
+using Pulumirpc;
+
+namespace Pulumi
+{
+    /// <summary>
+    /// Abstraction for hooks to the engine and monitor.
+    /// </summary>
+    internal interface IMonitor
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Deployment/IMonitor.cs]: See above, i would call this somethin glike `IRpc` or `IRpcDispatcher`  Then you can have your `MockRpcDispatcher` and `GrpcDispatcher`. 

From a regular design perspective i somewhat like that as it makes it very clear in our system at which point we're calling out through rpc to some other component.

File: sdk/dotnet/Pulumi/Serialization/OutputCompletionSource.cs [hunk 9]
```diff
@@ -100,5 +99,10 @@ internal static class OutputCompletionSource
             Log.Debug("Fields to assign: " + JsonSerializer.Serialize(result.Keys), resource);
             return result.ToImmutable();
         }
+
+        private static bool IsOutputType(this Type propType)
+        {
+            return propType.IsConstructedGenericType && propType.IsGenericType && propType.GetGenericTypeDefinition() == typeof(Output<>);
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Serialization/OutputCompletionSource.cs]: do we need both the `IsConstructedGenericType ` and `IsgenericType` checks?

File: sdk/dotnet/Pulumi/Serialization/OutputCompletionSource.cs [hunk 10]
```diff
@@ -100,5 +99,10 @@ internal static class OutputCompletionSource
             Log.Debug("Fields to assign: " + JsonSerializer.Serialize(result.Keys), resource);
             return result.ToImmutable();
         }
+
+        private static bool IsOutputType(this Type propType)
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Serialization/OutputCompletionSource.cs]: private helper.  would prefer this **not** be an extension.

File: sdk/dotnet/Pulumi/Testing/IMocks.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
+using System.Collections.Immutable;
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Testing/IMocks.cs]: copyriht.

File: sdk/dotnet/Pulumi/Testing/IMocks.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+using System.Threading.Tasks;
+
+namespace Pulumi.Testing
+{
+    /// <summary>
+    /// Hooks to mock the engine that provide test doubles for offline unit testing of stacks.
+    /// </summary>
+    public interface IMocks
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Testing/IMocks.cs]: don't like the name.  but not sure what a better name would be

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs [hunk 13]
```diff
@@ -0,0 +1,123 @@
+using System.Collections.Generic;
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: copyright.

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+        public MockMonitor(IMocks mocks)
+        {
+            _mocks = mocks;
+        }
+
+        public async Task<InvokeResponse> InvokeAsync(InvokeRequest request)
+        {
+            var result = await _mocks.CallAsync(request.Tok, ToDictionary(request.Args), request.Provider);
+            return new InvokeResponse { Return = await SerializeAsync(result) };
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Testing/MockMonitor.cs]:
```suggestion
            return new InvokeResponse { Return = await SerializeAsync(result).ConfigureAwait(false) };
```

Comment (CyrusNajmabadi) [reply sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: applies to all `await`s in this file.

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+        }
+
+        public async Task<ReadResourceResponse> ReadResourceAsync(Resource resource, ReadResourceRequest request)
+        {
+            var (id, state) = await _mocks.NewResourceAsync(request.Type, request.Name,
+                ToDictionary(request.Properties), request.Provider, request.Id);
+            this.Resources.Add(resource);
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: this needs to be threadsafe.  (applies to other functions as well).

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+            {
+                Urn = NewUrn(request.Parent, request.Type, request.Name),
+                Properties = await SerializeAsync(state)
+            };
+        }
+
+        public async Task<RegisterResourceResponse> RegisterResourceAsync(Resource resource, RegisterResourceRequest request)
+        {
+            var (id, state) = await _mocks.NewResourceAsync(request.Type, request.Name, ToDictionary(request.Object),
+                request.Provider, request.ImportId);
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: feels weird that we go through the same mock entrypoint for both Read and Register.  I would expect tests to want ot know the difference.

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task LogAsync(LogRequest request)
+        {
+            if (request.Severity == LogSeverity.Error)
+            {
+                this.Errors.Add(request.Message);
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: needs lock.

File: sdk/dotnet/Pulumi/Serialization/OutputCompletionSource.cs (hunk 9 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Serialization/OutputCompletionSource.cs]: yes, I've hit a problem with the old code which I'll now try to recall...

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+            }
+            
+            return Task.CompletedTask;
+        }
+
+        public Task<SetRootResourceResponse> SetRootResourceAsync(SetRootResourceRequest request)
+        {
+            _rootResourceUrn = request.Urn;
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: make threadsafe, bail if we get mltiple sets of this.

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+            return Task.FromResult(new SetRootResourceResponse());
+        }
+
+        public Task<GetRootResourceResponse> GetRootResourceAsync(
+            GetRootResourceRequest request)
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: unwrap.

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+            => Task.FromResult(new GetRootResourceResponse { Urn = _rootResourceUrn });
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: lock, bail if we don't have a root yet?

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+                }
+            }
+            return builder.ToImmutable();
+        }
+
+        private async Task<Struct> SerializeAsync(object o)
+        {
+            var dict = (o as IDictionary<string, object>)?.ToImmutableDictionary()
+                       ?? await _serializer.SerializeAsync("", o) as ImmutableDictionary<string, object>
+                       ?? ImmutableDictionary<string, object>.Empty;
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: can we be more precise here, and throw if we get a value that we can't deal with?

File: sdk/dotnet/Pulumi/Testing/IMocks.cs [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+        /// <summary>
+        /// Invoked when the program needs to call a provider to load data (e.g., to retrieve an existing
+        /// resource).
+        /// </summary>
+        /// <param name="token">Function token.</param>
+        /// <param name="args">Dictionary of input arguments.</param>
+        /// <param name="provider">Provider.</param>
+        /// <returns>Invocation result, can be either a POCO or a dictionary bag.</returns>
+        Task<object> CallAsync(string token, ImmutableDictionary<string, object> args, string? provider);
+    }
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Testing/IMocks.cs]: can we have a Default impl of this with virtuals/delegates for all of these?  Users can then instantiate and only provide the helpers they need.  i.e. many tests will not need to provide an impl for `CallAsync`.  

If we also split out thins like ReadResource and RegisterResource, it will make it easier to only supply the functions we want.


(note: we'd still have the interface, just also have a useful concrete impl people could use).

File: sdk/dotnet/Pulumi.Tests/StackTests.cs [hunk 23]
```diff
@@ -30,10 +28,14 @@ public ValidStack()
         [Fact]
         public async Task ValidStackInstantiationSucceeds()
         {
-            var (stack, outputs) = await Run<ValidStack>();
-            Assert.Equal(2, outputs.Count);
-            Assert.Same(stack.ExplicitName, outputs["foo"]);
-            Assert.Same(stack.ImplicitName, outputs["ImplicitName"]);
+            var result = await Deployment.TestAsync<ValidStack>(new DefaultMocks());
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi.Tests/StackTests.cs]: where is DefaultMocks defined?  i didn't see it.

File: sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+        }
+
+        public async Task<InvokeResponse> InvokeAsync(InvokeRequest request)
+            => await this._client.InvokeAsync(request);
+
+        public async Task<ReadResourceResponse> ReadResourceAsync(Resource resource, ReadResourceRequest request)
+            => await this._client.ReadResourceAsync(request);
+
+        public async Task<RegisterResourceResponse> RegisterResourceAsync(Resource resource, RegisterResourceRequest request)
+            => await this._client.RegisterResourceAsync(request);
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs]: you don't use the `Resource resource` in the body.  why was it necessary?

Comment (CyrusNajmabadi) [reply sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs]: i see.  you use it in tests.   don't love it.

Might consider that if we have our own IRpcDispatcher interface, we have a few classes/structs we define that are independent of grpc.  It could hold onto the data and the resource.  the Grpc impl would then create the grpc type, the test version could just check the data however it wants.

Not certain if this is worth it... but it's something i wouldn't mind us looking into.

File: sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+            => await this._client.RegisterResourceOutputsAsync(request);
+
+        public async Task LogAsync(LogRequest request)
+            => await this._engine.LogAsync(request);
+
+        public async Task<SetRootResourceResponse> SetRootResourceAsync(SetRootResourceRequest request)
+            => await this._engine.SetRootResourceAsync(request);
+
+        public async Task<GetRootResourceResponse> GetRootResourceAsync(GetRootResourceRequest request)
+            => await this._engine.GetRootResourceAsync(request);
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs]: can remove async/await from all of these methods.

**Comment by CyrusNajmabadi** [comment]
Overall this is very good and looks to eb the right way forward.  Most of my comments are about cleanup/ergonomics.  But this is overall great!

File: sdk/dotnet/Pulumi/Deployment/Deployment_RootResource.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply sdk/dotnet/Pulumi/Deployment/Deployment_RootResource.cs]: Either `IDispatcher` or `IRpcDispatcher`.  I actually like the latter as it strongly conveys "this is where call out to some other component elsewhere".  It also has good impl names.  i.e. GrpcDispatcher and MockRpcDispatcher.  Both very clear in terms of what they're responsible for.

File: sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs (hunk 24 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs]: yeah, it's an ugly way to build up a list of resources for tests

File: sdk/dotnet/Pulumi/Testing/IMocks.cs (hunk 22 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/IMocks.cs]: I think most users would use something like `Moq` to create mocks, and then they don't have to mock methods that aren't needed.
We still may need a default implementation, but I haven't spent time thinking it through yet, so it's in the Test project for now.
re `ReadResource`/`RegisterResource` -- I tried to keep it in line with JS in https://github.com/pulumi/pulumi/pull/3738

File: sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs (hunk 25 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs]: ah, missed this when got this code from IDE auto-refactoring. any idea why they do it by default?

File: sdk/dotnet/Pulumi/Testing/IMocks.cs (hunk 22 above)
Comment (CyrusNajmabadi) [reply sdk/dotnet/Pulumi/Testing/IMocks.cs]: rgd `Moq`.  That's a good point.  In that case, i would like us to do the same so people can look at our tests for how to do things.  That can come in a followup PR though.

> I tried to keep it in line with JS in #3738

Hrmm.. def eems odd to me. @pgavlin could you give some info on why you thought this was the right way to go here?

Comment (CyrusNajmabadi) [reply sdk/dotnet/Pulumi/Testing/IMocks.cs]: Although... thinking about it more... i can see why it would be done this way.  This is hte high level SDK concept someone is using.  And if there's an 'id' or not lets people know if it was a read or a register... i could get behind this :)

Comment (pgavlin) [reply sdk/dotnet/Pulumi/Testing/IMocks.cs]: > could you give some info on why you thought this was the right way to go here?

Can you be more specific? I'm not sure what you're asking about.

If you're asking about the design of the mocks interface itself, I chose the `newResource` approach because in most scenarios, it doesn't matter whether or not a resource is being created or read: in either case, a user must return some sort of valid state for the resource.

File: sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs (hunk 25 above)
Comment (CyrusNajmabadi) [reply sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs]: which refactoring were you doing?

File: sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs (hunk 4 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs]: Gotcha. I changed it because we return `Task.CompletedTask` from mocks and they obviously collide here.
`Task.CompletedTask` is not guaranteed to be the same instance, so we can't really compare to it. Would it be safe to ignore completed tasks here?

File: sdk/dotnet/Pulumi/Serialization/OutputCompletionSource.cs (hunk 9 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Serialization/OutputCompletionSource.cs]: no, we don't :) I reverted the change except I fixed the original conditional

File: sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs (hunk 25 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/GrpcMonitor.cs]: It's actually more tricky: if I remove async-await, I get
`Cannot convert expression type 'Grpc.Core.AsyncUnaryCall<Pulumirpc.InvokeResponse>' to return type 'System.Threading.Tasks.Task<Pulumirpc.InvokeResponse>'`
So, `await` "converts" `AsyncUnaryCall` to `Task`.

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs (hunk 16 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: Discussed in another thread, resolving this one

File: sdk/dotnet/Pulumi/Testing/IMocks.cs (hunk 12 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/IMocks.cs]: Its sole usage is being an argument of `TestAsync` method, so mocks is what it is. We can go for `IEngine`, `IRemoteCalls`, ... and keep `mocks` for the parameter name

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs (hunk 15 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: added a lock around it

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs (hunk 17 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: added

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs (hunk 18 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: added

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs (hunk 20 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: added

File: sdk/dotnet/Pulumi/Testing/MockMonitor.cs (hunk 21 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/MockMonitor.cs]: changed

File: sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs (hunk 4 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs]: Fixed by merging the master

File: sdk/dotnet/Pulumi/Testing/IMocks.cs (hunk 12 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/IMocks.cs]: This aligns with the naming in TS/Python, so I'll leave it at `IMocks`.

File: sdk/dotnet/Pulumi/Deployment/Deployment_Run.cs [hunk 26]
```diff
@@ -87,6 +89,32 @@ public static Task<int> RunAsync(Func<Task<IDictionary<string, object?>>> func)
         public static Task<int> RunAsync<TStack>() where TStack : Stack, new()
             => CreateRunner().RunAsync<TStack>();
 
+        /// <summary>
+        /// Entry point to test a Pulumi application. Deployment will
+        /// instantiate a new stack instance based on the type passed as TStack
+        /// type parameter. This method creates no real resources.
+        /// Note: Currently, unit tests that call <see cref="TestAsync{TStack}"/>
+        /// must run sequentially; parallel execution is not supported.
```
Comment (lukehoban) [inline sdk/dotnet/Pulumi/Deployment/Deployment_Run.cs]: Do we have any path to enabling this?  It does seem a likely very significant restriction. 

Also - if this has to be the case for now - should we at least enforce it with a good error message?

File: sdk/dotnet/Pulumi/Serialization/OutputCompletionSource.cs [hunk 27]
```diff
@@ -75,7 +75,7 @@ internal static class OutputCompletionSource
             {
                 var propType = prop.PropertyType;
                 var propFullName = $"[Output] {resource.GetType().FullName}.{prop.Name}";
-                if (!propType.IsConstructedGenericType &&
+                if (!propType.IsConstructedGenericType ||
```
Comment (lukehoban) [inline sdk/dotnet/Pulumi/Serialization/OutputCompletionSource.cs]: Is this intentional?  It doesn’t look right.

File: sdk/dotnet/Pulumi/Testing/TestResult.cs [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+        /// </summary>
+        public ImmutableDictionary<string, object> StackOutputs { get; }
+
+        /// <summary>
+        /// Extract a value from an <see cref="Output{T}"/>. Can only be used
+        /// for outputs fount in <see cref="Resources"/>.
+        /// </summary>
+        /// <param name="output">An output containing a value.</param>
+        /// <returns>An awaitable task that returns the value.</returns>
+        public Task<T> GetAsync<T>(Output<T> output) => output.GetValueAsync();
```
Comment (lukehoban) [inline sdk/dotnet/Pulumi/Testing/TestResult.cs]: Why do we need/want to expose this here?

File: sdk/dotnet/Pulumi/Deployment/Deployment.cs [hunk 29]
```diff
@@ -93,27 +89,32 @@ private Deployment()
...
+            _logger = new Logger(this, this.RpcDispatcher);
+        }
 
+        /// <summary>
+        /// This constructor is called from <see cref="TestAsync{TStack}"/> with
+        /// a mocked monitor and dummy values for project and stack.
+        /// </summary>
+        private Deployment(IRpcDispatcher rpcDispatcher)
+        {
+            _isDryRun = true;
```
Comment (lukehoban) [inline sdk/dotnet/Pulumi/Deployment/Deployment.cs]: Why is isDryRun set to true here?

File: sdk/dotnet/Pulumi/Deployment/Deployment.cs [hunk 30]
```diff
@@ -93,27 +89,32 @@ private Deployment()
...
+            _stackName = "teststack";
+            _projectName = "testproject";
```
Comment (lukehoban) [inline sdk/dotnet/Pulumi/Deployment/Deployment.cs]: Feels a little odd to set these to test names at this level. Also - these names seem different than in https://github.com/pulumi/pulumi/pull/3738.

File: sdk/dotnet/Pulumi/Deployment/GrpcDispatcher.cs [hunk 31]
```diff
@@ -0,0 +1,41 @@
+// Copyright 2016-2020, Pulumi Corporation
+
+using System.Threading.Tasks;
+using Grpc.Core;
+using Pulumirpc;
+
+namespace Pulumi
+{
+    internal class GrpcDispatcher : IRpcDispatcher
```
Comment (lukehoban) [inline sdk/dotnet/Pulumi/Deployment/GrpcDispatcher.cs]: Why did we have to introduce this combined gRPC interface here but not in https://github.com/pulumi/pulumi/pull/3738?

File: sdk/dotnet/Pulumi/Serialization/OutputCompletionSource.cs (hunk 27 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Serialization/OutputCompletionSource.cs]: This is an intentional fix for a bug which got exposed while testing this PR. The idea is `not(a & b)` so it should be (not a) or (not b)

File: sdk/dotnet/Pulumi/Deployment/Deployment_Run.cs (hunk 26 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/Deployment_Run.cs]: Not really... the whole .NET SDK implementation relies heavily on static state, which means parallel executions would bite each other.

We had
```
#3399 (comment). Can we get off of using static state. Perhaps have the instance
state computed and passed into the RunAsync method as an appropriate context.
Note: inputs/outputs are difficult here as they look at state like IsPreview
and don't always have a clear way to determine that value.
```
in #3470 but it was never considered a priority.

Enforcing sounds like a good idea.

File: sdk/dotnet/Pulumi/Testing/TestResult.cs (hunk 28 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/TestResult.cs]: Basically, to assert output values. E.g.
```cs
var consumptionPlan = result.Resources.OfType<Plan>().Single();
var kind = await result.GetAsync(consumptionPlan.Kind);
kind.Should().Be("FunctionApp");
```
Do you have better ideas?

File: sdk/dotnet/Pulumi/Deployment/Deployment.cs (hunk 29 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/Deployment.cs]: A unit test is like a preview, isn't it?

File: sdk/dotnet/Pulumi/Deployment/Deployment.cs (hunk 30 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/Deployment.cs]: What would be the right level to set them?
(I'll change it to `stack` and `project`.)

File: sdk/dotnet/Pulumi/Deployment/GrpcDispatcher.cs (hunk 31 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/GrpcDispatcher.cs]: We don't have to... we could introduce two separate interfaces instead. But that's what we eventually came up with Cyrus.

Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/GrpcDispatcher.cs]: On a second look, the reason to merge seems arbitrary. I reverted the change and kept the `Engine` and `Monitor` entities separate.

File: sdk/dotnet/Pulumi/Deployment/Deployment.cs (hunk 30 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/Deployment.cs]: Changed to `stack` and `project`.

File: sdk/dotnet/Pulumi/Deployment/Deployment_Run.cs (hunk 26 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/Deployment_Run.cs]: Added the checks around `TestAsync` to throw an exception if multiple parallel executions are detected

File: sdk/dotnet/Pulumi/Testing/TestResult.cs (hunk 28 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Testing/TestResult.cs]: Removed this to move the PR to completion. Need to refine the pattern while writing docs.

File: sdk/dotnet/Pulumi/Deployment/Deployment.cs (hunk 29 above)
Comment (pgavlin) [reply sdk/dotnet/Pulumi/Deployment/Deployment.cs]: Not necessarily: at least in other SDKs, the behavior of `apply` and some of the marshalling code differs in a preview. I think we want this to be configurable, as it is in other SDKs.

Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/Deployment.cs]: @pgavlin Good point! I introduced options for preview, project, and stack akin to other SDKs.
